### PR TITLE
RavenDB-4608 Static indexes

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs [hunk 1]
```diff
@@ -213,51 +287,93 @@ private static void AddValueToReduceHash(object value, Hashing.Streamed.XXHash64
...
+                                break;
+                            case Mode.MultipleValues:
+                                CopyToBuffer(lcsv.Buffer, lcsv.CompressedSize);
+                                break;
+                        }
+
                         return;
                     }
 
                     if (value is long)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs]: need to also handle complex objects

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs [hunk 2]
```diff
@@ -106,7 +106,7 @@ private void CreateIndexWriter()
             using (indexWriter.MergeScheduler)
             {
             }
-            indexWriter.SetMergeScheduler(new ErrorLoggingConcurrentMergeScheduler());
+            indexWriter.SetMergeScheduler(new SerialMergeScheduler());
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:109]: What is the reasoning here?

Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:109]: ConcurrentMergeScheduler spawns threads for merging and we:
1) have only one write transaction that would be accesed concurrently by those threads
2) using ThreadLocal inside lucene voron directory

I will add a TODO to my list to check what we can do about this, but I was getting NRE here due to ThreadLocal usage

File: src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs (hunk 1 above)
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs]: Yes. I'm working on support for complex / nested objects, generally in indexes, not only here. 
In addition to the above, we need to take care of:
- storing in Lucene - probably write as JSON to allow search on them (such field can be available for full text search)
- implement `Equals` for blittable / dynamic blittable objects to properly compare i.e. GroupBy a complex object
- handling nested objects from Lucene index on querying

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs (hunk 2 above)
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:109]: I'm actually fine with it being serial, we just need to be able to track when this is happening and add it to the metrics.

**Pull request merged**